### PR TITLE
Create separate arguments data and ppp version

### PR DESCRIPTION
**Pull request by shahronak47**
- Added new arguments called `release_version`, `ppp_version` and `identity` in the `pip` api call. 
- Added supporting functions `return_correct_version`, `select_max_version_from_ppp` and `select_max_version_from_release_version` write corresponding units tests for it. 
- Using the supporting functions fill the `params$version` argument so that rest of the API is not affected.

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -496,3 +496,30 @@ clear_cache <- function(cd) {
...
+#' @return character
+#'
+#' @noRd
+#'
+create_folder_id <- function(data_version , ppp_version, versions_available) {
+  #From the available versions subset the ones for this data and ppp versions
+  selected_version <- grep(sprintf('^%s_%s', data_version, ppp_version), versions_available, value = TRUE)
+  if(length(selected_version) == 0)
+    #Since the function returns character values
+    return("404")
```
Comment (shahronak47) [inline R/utils.R:515]: If there is no version match found I am  returning a character "404". This works but I am not sure if this is the best return value here.

File: inst/plumber/v1/endpoints.R [hunk 2]
```diff
@@ -281,22 +276,22 @@ function(req) {
 #* @param welfare_type:[chr] Welfare Type. Options are "income" or "consumption"
 #* @param reporting_level:[chr] Reporting level. Options are "national", "urban", "rural".
 #* @param ppp:[dbl] Custom Purchase Power Parity (PPP) value.
-#* @param version:[chr] Data version. Defaults to most recent version. See api/v1/versions for all available versions
+#* @param data_version:[chr] date when the data was published in YYYYMMDD format
+#* @param ppp_version:[chr] ppp year to be used
```
Comment (shahronak47) [inline inst/plumber/v1/endpoints.R:280]: Both the new arguments I have kept it as character since year is a character value as well.

File: inst/plumber/v1/endpoints.R [hunk 3]
```diff
@@ -281,22 +276,22 @@ function(req) {
...
 
   # Define default arguments
-  # if (is.null(params$country))
-  #   params$country <- "all"
-  # if (is.null(params$year))
-  #   params$year <- "all"
+  if (is.null(params$country))
+    params$country <- "all"
+  if (is.null(params$year))
+    params$year <- "all"
```
Comment (shahronak47) [inline inst/plumber/v1/endpoints.R:294]: I have uncommented these lines so the default `pip` API call works without the `country` and `year` argument.  I was curious why was this commented in the first place? Mostly, I am not aware of a certain scenario here.

File: inst/plumber/v1/endpoints.R [hunk 4]
```diff
@@ -9,23 +9,23 @@ library(pipapi)
...
-  if (!is.null(req$argsQuery$version) & !grepl("swagger", req$PATH_INFO)) {
-    if (!is.null(req$argsQuery$version)) {
-      if (!req$argsQuery$version %in% lkups$versions) {
+
+  #If no arguments are passed, use the latest version
+  if (is.null(req$argsQuery$release_version) && is.null(req$argsQuery$ppp_version) &&
+     is.null(req$argsQuery$version) && is.null(req$argsQuery$identity)) {
+      version <- lkups$latest_release
+  } else {
+    version <- pipapi::return_correct_version(req$argsQuery$version, req$argsQuery$release_version, req$argsQuery$ppp_version, req$argsQuery$identity, lkups$versions)
```
Comment (tonyfujs) [inline inst/plumber/v1/endpoints.R:18]: I exported the function, otherwise it is not being found when the API is running

File: inst/plumber/v1/endpoints.R [hunk 5]
```diff
@@ -9,23 +9,23 @@ library(pipapi)
...
-  # tictoc::tic("filters")
-  # browser()
-  if (!is.null(req$argsQuery$version) & !grepl("swagger", req$PATH_INFO)) {
-    if (!is.null(req$argsQuery$version)) {
-      if (!req$argsQuery$version %in% lkups$versions) {
+
+  #If no arguments are passed, use the latest version
+  if (is.null(req$argsQuery$release_version) && is.null(req$argsQuery$ppp_version) &&
+     is.null(req$argsQuery$version) && is.null(req$argsQuery$identity)) {
+      version <- lkups$latest_release
```
Comment (tonyfujs) [inline inst/plumber/v1/endpoints.R:16]: I think we should use single `&` here. THe `&&` silently drops extra items if length is > to 1. We should not run in such cases, but I guess it's always better to get a warning.

File: R/utils-plumber.R [hunk 6]
```diff
@@ -222,3 +222,60 @@ extract_endpoint <- function(path) {
...
+  if (!is.null(release_version) & !is.null(ppp_version) & !is.null(identity)) {
+    selected_version <- grep(sprintf('^%s_%s_\\d{2}_\\d{2}_%s$', release_version, ppp_version, identity), versions_available, value = TRUE)
+  } else if (!is.null(release_version) & !is.null(ppp_version)) {
+    # This probably would never be executed since identity would never be NULL.
+    selected_version <- grep(sprintf('^%s_%s_\\d{2}_\\d{2}_[A-Z]+$', release_version, ppp_version), versions_available, value = TRUE)
+  } else if (!is.null(release_version) & !is.null(identity)) {
+    selected_version <- grep(sprintf('^%s_\\d{4}_\\d{2}_\\d{2}_%s$', release_version, identity), versions_available, value = TRUE)
+  } else if (!is.null(ppp_version) & !is.null(identity)) {
+    selected_version <- grep(sprintf('\\d{6}_%s_\\d{2}_\\d{2}_%s$', ppp_version, identity), versions_available, value = TRUE)
+  }
```
Comment (tonyfujs) [inline R/utils-plumber.R:256]: Would it make sense to create sub-functions that we can unit test here? I'm always a bit nervous when using `regex`...

File: inst/plumber/v1/endpoints.R (hunk 3 above)
Comment (tonyfujs) [reply inst/plumber/v1/endpoints.R:296]: The default argument values are handled at the `pip()` function level, so I don't think these lines are necessary.

Comment (tonyfujs) [reply inst/plumber/v1/endpoints.R:296]: Oh I see... This is needed for the parallel processing below. Mmm... I don't remember exactly why this has been commented out, but I think this is taken care of in one of the filters, so we should never get NULL values ofr country and year at this stage. To be checked.

**Review by tonyfujs — CHANGES_REQUESTED** [review]

**Comment by tonyfujs** [comment]
Thanks @shahronak47 
It looks good, a few changes before we merge:
* The new parameters should be documented in the Open API yml file
* The `/versions` endpoint currently returns the full version IDs only, it should also return the available `release_versions` and `pp_versions` now.
* A couple of edge cases. Not sure how to handle them, but the current behavior is not ideal...
  *  If a specific release date is only available for indentity=INT, then specifying only this release date in the query will return an error without any informative message.
  * Same issue happens if we select only ppp_version=2017 (since the default for idendity is "PROD")

File: inst/plumber/v1/endpoints.R (hunk 5 above)
Comment (shahronak47) [reply inst/plumber/v1/endpoints.R:16]: That should not be an issue moving forward. In R, 4.2 length > 1, returns a warning which eventually will start returning error since 4.2.1. 

https://cran.r-project.org/doc/manuals/r-devel/NEWS.html 

>CHANGES IN R 4.2.0
Calling && or || with either argument of length greater than one now gives a warning (which it is intended will become an error).

>CHANGES IN R-devel
Calling && or || with LHS or (if evaluated) RHS of length greater than one is now always an error, with a report of the form

    'length = 4' in coercion to 'logical(1)'

Moreover, `&&` short-circuits the expression and doesn't evaluate other conditions if any of earlier conditions is `FALSE` making it faster than `&`. 

Let me know if you still want me to switch it back to `&` in case if we are not going to use R 4.2 or higher.

File: R/utils-plumber.R (hunk 6 above)
Comment (shahronak47) [reply R/utils-plumber.R]: yes, that makes sense. I have done so and committed the change. However, I was hoping to find a solution which did not involve so many `if...else` statements but unable to come up with that logic. So we will keep this for now.

**Comment by shahronak47** [comment]
Ok. I have done the requested changes. 

>The /versions endpoint currently returns the full version IDs only, it should also return the available release_versions and pp_versions now.

The end point output has been changed to return a named list now. Should there be any change for `get_pip_version` function as well ?

>If a specific release date is only available for indentity=INT, then specifying only this release date in the query will return an error without any informative message.

Not sure, if I understand this clearly. I have `lkups$versions` as 

```
lkups$versions
[1] "20220609_2011_02_02_PROD" "20220504_2017_01_02_PROD" "20211212_2011_01_01_PROD"
[4] "20200101_2011_01_01_PROD" "20220602_2017_01_02_INT"  "20220504_2017_01_02_INT" 
```

`release_version = "20220602"` is only available for `INT` and 

```
 return_correct_version(release_version = "20220602", versions_available = lkups$versions)
[1] "404"
```
returns 404 correctly? since `identity` is by default taken as `"PROD"` ?

**Comment by tonyfujs** [comment]
> Ok. I have done the requested changes.
> 
> > The /versions endpoint currently returns the full version IDs only, it should also return the available release_versions and pp_versions now.
> 
> The end point output has been changed to return a named list now. Should there be any change for `get_pip_version` function as well ?

Thanks. This should be updated so the response is a data.frame otherwise Stata users won't have access to this endpoints since Stata does not have a json parser...

File: R/utils-plumber.R [hunk 7]
```diff
@@ -222,3 +222,124 @@ extract_endpoint <- function(path) {
...
+  if (length(selected_version) == 0)
+    #Since the function returns character values
+    return("404")
+  #If only 1 value matches
+  else if (length(selected_version) == 1)
+    return(selected_version)
+  #If more than 1 value matches
+  #If release_version is not null get the max version from ppp date
+  if (!is.null(release_version)) return(select_max_version_from_ppp(selected_version))
+  else return(select_max_version_from_release_version(selected_version))
```
Comment (randrescastaneda) [inline R/utils-plumber.R:267]: If I remember correctly, the versioning nomenclature works in a way such that the following code always returns the correct version. In other words, there is no need for this "if-else" conditions. Could you please confirm? 

Thanks. 

```r
selected_version[which.max(selected_version)]
```

Comment (shahronak47) [reply R/utils-plumber.R:267]: Maybe you are right. However, `selected_version` is a string (Eg - `20220609_2011_02_02_PROD`) and `which.max` is not recommended to use on string values.

Comment (randrescastaneda) [reply R/utils-plumber.R:267]: Fair enough. I actually don't use `which.max` but `max`, which works well with characters. So, this is more than enough, `max(selected_version)`. Since the characters we use in the versioning are relatively standard, and do not depend on any particular locale,  I think we can use `max` safely. Please, try it out and let me know how it goes. 

Thanks!

Comment (shahronak47) [reply R/utils-plumber.R]: I thought about this and test on few examples and I think you are right. Using only `max` should be sufficient here. I made the changes accordingly. Thank you for the feedback.

Comment (tonyfujs) [reply R/utils-plumber.R]: Hey - What about this:
```
tmp <- c("20221212_2017_03_01_PROD", "20221212_2017_03_01_TEST")
> max(tmp)
[1] "20221212_2017_03_01_TEST"
```

It looks like the suffix can be problematic for returning the correct version here, no? If a *_TEST folder makes it to the PROD server by mistake, this could be very problematic. What do you think?

Even if the `select_max_version_from_release_version` function is just a thin wrapper around `max(selected_version)`, I would rather have it as a separate function so we can unit test it for such edge cases. I'm reviewing this PR now, and will add the function back along with it's unit tests.

Comment (shahronak47) [reply R/utils-plumber.R]: but I can't think of scenario where we will have version from more than one `identity` in `selected_version`. We always have an `identity` argument with default value as `PROD` so even if `_TEST` folder makes it to PROD it will never be selected unless you explicitly specify `identity=TEST`.

File: inst/plumber/v1/endpoints.R [hunk 8]
```diff
@@ -281,22 +270,24 @@ function(req) {
...
 
   # Define default arguments
-  # if (is.null(params$country))
-  #   params$country <- "all"
-  # if (is.null(params$year))
-  #   params$year <- "all"
+  if (is.null(params$country))
+    params$country <- "all"
+  if (is.null(params$year))
+    params$year <- "all"
```
Comment (tonyfujs) [inline inst/plumber/v1/endpoints.R:290]: Ok, so this handled in one of the filters and can therefore be removed.
image

File: inst/plumber/v1/endpoints.R [hunk 9]
```diff
@@ -9,23 +9,25 @@ library(pipapi)
...
-  # tictoc::tic("filters")
-  # browser()
-  if (!is.null(req$argsQuery$version) & !grepl("swagger", req$PATH_INFO)) {
-    if (!is.null(req$argsQuery$version)) {
-      if (!req$argsQuery$version %in% lkups$versions) {
+
+  #If no arguments are passed, use the latest version
+  if (is.null(req$argsQuery$release_version) && is.null(req$argsQuery$ppp_version) &&
+     is.null(req$argsQuery$version) && is.null(req$argsQuery$identity)) {
+      version <- lkups$latest_release
```
Comment (tonyfujs) [inline inst/plumber/v1/endpoints.R:16]: I have finally replaced the `&&` by `&`. We are still using 4.0.2 in PROD (soon to be updated to 4.1.2) and `&` is significantly faster than `&&` (even if we are in nanoseconds territory)

File: inst/plumber/v1/endpoints.R [hunk 10]
```diff
@@ -9,23 +9,25 @@ library(pipapi)
...
+  } else {
+    if(is.null(req$argsQuery$identity)) req$argsQuery$identity <- 'PROD'
+    version <- pipapi::return_correct_version(req$argsQuery$version, req$argsQuery$release_version, req$argsQuery$ppp_version, req$argsQuery$identity, lkups$versions)
+  }
+    #If the version is not found (404) or it is not present in valid versions vector return an error.
+    if (version == "404" || !version %in% lkups$versions) {
```
Comment (tonyfujs) [inline inst/plumber/v1/endpoints.R:22]: I removed `version == "404"` and left only `if (!version %in% lkups$versions)`, which should be sufficient

**Review by tonyfujs — APPROVED** [review]
Thanks @shahronak47 
I made a couple of slight changes that I will push now, and then merge to `dev`
